pkg/datapath/linux/node_ids.go

```diff
@@ -106,10 +106,21 @@ func (n *linuxNodeHandler) allocateIDForNode(oldNode *nodeTypes.Node, node *node
 
 	for _, addr := range node.IPAddresses {
 		ip := addr.IP.String()
-		if _, exists := n.nodeIDsByIPs[ip]; exists {
+		if id, exists := n.nodeIDsByIPs[ip]; exists && id == nodeID {
 			if !SPIChanged {
 				continue
 			}
+		} else if exists && id != nodeID {
```

Comment:
I tried to explain how this can occur in the commit message, but I'll try again here.

The prerequisites are as follows:

1. Nodes have multiple IPs, though not all of them might be available on initial update.
2. NodeIDs are per-node state. Each node assigns NodeIDs to other nodes, which likely don't match with NodeIDs assigned by other nodes.
3. The cilium agent can experience downtime, a common case is during upgrade.
4. While the cilium agent is down, nodes can be removed from the cluster.
5. When nodes are added to the cluster, they may reuse IPs previously used by other nodes.

Now, if cilium agent on node A is down while node B `{1.1.1.1, 10.1.1.1, nodeID: 2}` is removed from the cluster, it will _not_ clean up the mappings of B's IPs. If then node C joins the cluster and gets `1.1.2.3` initially, and in a later update also `10.1.1.1`, we have a mismatch. `1.1.2.3` looks like an IP we have never seen and allocate a new nodeID for (say, 3). `10.1.1.1` on the other hand is known, and still maps to nodeID 2. Since `getNodeIDForNode` for some reason prefers IPs that are later in some slice order, we will pick `nodeID == 2` but `id == 3` when IP is `1.1.2.3`.

Reply:
Thanks @bimmlerd the key point I was missing here is that we reload the stale NodeIDs from our eBPF maps during agent restart. I was thinking we generate new NodeIDs on every restart. 

This change now makes sense to me.